### PR TITLE
Transition to Typescript

### PR DESCRIPTION
**Pull request by PopGoesTheWza**
Hi,

I am experimenting with this package and felt like contributing a little. Here's a first draft at switching to Typescript.

Regards,

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -14,8 +14,8 @@ jobs:
           - 18
           - 16
```
Comment (PopGoesTheWza) [inline .github/workflows/main.yml:15]: Should support of node 16 be dropped?

File: tsconfig.json [hunk 2]
```diff
@@ -1,12 +1,14 @@
 {
+	"extends": "@sindresorhus/tsconfig",
 	"compilerOptions": {
 		"lib": [
 			"es2021"
```
Comment (PopGoesTheWza) [inline tsconfig.json:5]: Could be `es2023` if support of node 16 is dropped

File: source/index.ts [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+	'constructor',
+]);
+
+const digits = new Set('0123456789');
+
+// eslint-disable-next-line complexity
+function getPathSegments(path: string) {
+	const parts = [];
+	let currentSegment = '';
+	let currentPart = 'start';
```
Comment (PopGoesTheWza) [inline source/index.ts:28]: Idea: use `Symbol`s in place of string constants

File: source/index.ts [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+getProperty({foo: {'dot.dot': 'unicorn'}}, 'foo.dot\\.dot');
+//=> 'unicorn'
+
+getProperty({foo: [{bar: 'unicorn'}]}, 'foo[0].bar');
+//=> 'unicorn'
+```
+*/
+export function getProperty<ObjectType, PathType extends string, DefaultValue = undefined, ReturnType = ObjectType extends AnObject | AFunction ? (unknown extends Get<ObjectType, PathType> ? DefaultValue : Get<ObjectType, PathType>) : undefined>(
+	object: ObjectType,
+	path?: PathType,
```
Comment (PopGoesTheWza) [inline source/index.ts:218]: Fix: original typescript definition had `path` as manadatory parameter

File: source/index.ts [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+			// but we want `get({foo: null}, 'foo.bar')` to equal `undefined`, or the supplied value, not `null`
+			if (index !== pathArray.length - 1) {
+				return defaultValue as ReturnType;
+			}
+
+			break;
+		}
+	}
+
+	return (object === undefined ? defaultValue : object) as ReturnType;
```
Comment (PopGoesTheWza) [inline source/index.ts:251]: Generic type `ReturnType` may need more work. Currently compatible with original typescript definition but not always accurate.

**Comment by sindresorhus** [comment]
While I appreciate the pull request, it's generally better to open an issue before doing large changes like this. I'm not really interested in moving to TypeScript. It adds a lot of development overhead without much gain. You gain most with TypeScript from codebases that are constantly changing, but this package is pretty much done.

**Pull request closed without merging**